crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -216,74 +216,139 @@ impl<'db> InnerIntersectionBuilder<'db> {
     }
 
     /// Adds a positive type to this intersection.
-    fn add_positive(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+    fn add_positive(&mut self, db: &'db dyn Db, new_positive: Type<'db>) {
         // TODO `Any`/`Unknown`/`Todo` actually should not self-cancel
-        match ty {
-            Type::Intersection(inter) => {
-                let pos = inter.positive(db);
-                let neg = inter.negative(db);
-                self.positive.extend(pos.difference(&self.negative));
-                self.negative.extend(neg.difference(&self.positive));
-                self.positive.retain(|elem| !neg.contains(elem));
-                self.negative.retain(|elem| !pos.contains(elem));
+        if let Type::Intersection(other) = new_positive {
+            for pos in other.positive(db) {
+                self.add_positive(db, *pos);
             }
-            _ => {
-                if !self.negative.remove(&ty) {
-                    self.positive.insert(ty);
-                };
+            for neg in other.negative(db) {
+                self.add_negative(db, *neg);
+            }
+        } else {
+            // ~Literal[True] & bool = Literal[False]
+            if let Type::Instance(class_type) = new_positive {
+                if class_type.is_known(db, KnownClass::Bool) {
+                    let mut found_bool_literal = None;
+                    for neg in &self.negative {
+                        if let Type::BooleanLiteral(bool) = neg {
+                            found_bool_literal = Some(*bool);
+                        }
+                    }
+
+                    if let Some(bool) = found_bool_literal {
+                        *self = Self::new();
+                        self.positive.insert(Type::BooleanLiteral(!bool));
+                        return;
+                    }
+                }
+            }
+
+            let mut to_remove = None;
+            for (index, existing_positive) in self.positive.iter().enumerate() {
+                // S & T = S    if S <: T
+                if existing_positive.is_subtype_of(db, new_positive) {
+                    return;
+                }
+                // same rule, reverse order
+                if new_positive.is_subtype_of(db, *existing_positive) {
+                    to_remove = Some(index);
```

Comment:
Nit: add break?

Reply:
To be honest, I'd rather not `break` here. I feel it adds a lot of mental overhead and has questionable benefit.

If we add a `break`, we need to make sure that other simplifications in the same loop (right now, only the *"A & B = Never, if A and B are disjoint"* rule below) can not trigger *in addition* to this one — they might be more powerful. Right now, the only counter-example I can think of would be something like `(int & ~Literal[1]) & Never`. When we add `Never` as a third element to the intersection, this `new_positive <: existing_positive` rule triggers first, because `Never <: int`, and would lead to a removal of `int`. However, if we keep going, we would also find out that `Never` and `int` are also disjoint! This would not just result in a `break`, but in a `return`, because we can immediately conclude that the overall result will be `Never`. We would not just skip this loop, but also the "negative" loop below.

I tried to write a test-case to show that `break`ing here is wrong, and eventually gave up. The example from above works in both cases (with and without `break`) because there are other simplifications in the `negative` branch below. But I'm not sure it invalidates my argument from above. I find it much easier to reason about this code (in the future, if we add more simplifications) if we don't add additional `break`s.

Reply:
I think it would be good in that case to add a comment why we aren't breaking early. I could see myself come by this code and add the break and then test if there are any failing tests.

Reply:
I first added the comment, but now removed it again. We now support removing multiple elements, so we can't break.